File: workspaces/001.fundamentals/code/02.order_service.rb
class OrderServiceRefactored

  # GOOD: The signature reads like a sentence.
  # Logic is delegated to the objects.
  def place_order(customer, address, pizza)

    # Step 1: Logic is calculated by the expert (the Pizza object)
    price = pizza.calculate_price

    # Step 2: Printing is simplified using methods on the objects
    puts "Order for: #{customer.contact_info}"
    puts "Ship to: #{address}" # Uses address.to_s
    puts "Total: $#{price}"
  end
end

# Encapsulates customer validation
class Customer
  attr_reader :name, :email

  def initialize(name, email)
    raise ArgumentError, "Invalid email" unless email.include?("@")
    @name = name
    @email = email
  end

  def contact_info
    "#{name} (#{email})"
  end
end

# Encapsulates address formatting
class Address
  attr_reader :street, :city, :zip_code

  def initialize(street, city, zip_code)
    raise ArgumentError, "Invalid zip" unless zip_code.length == 5
    @street = street
    @city = city
    @zip_code = zip_code
  end

  def to_s
    "#{street}, #{city} #{zip_code}"
  end
end

# Encapsulates pricing logic
class Pizza
  attr_reader :size, :toppings_count

  def initialize(size, toppings_count)
    @size = size
    @toppings_count = toppings_count
  end

  def calculate_price
    base_price = case size
                 when "Small" then 10.00
                 when "Medium" then 12.00
                 when "Large" then 14.00
                 else 0.0
                 end
    base_price + (toppings_count * 1.50)
  end
end

File: workspaces/001.fundamentals/code/02.order_service.java
public class OrderServiceRefactored {

    // GOOD: The signature reads like a sentence.
    // Logic is delegated to the objects.
    public void placeOrder(Customer customer, Address address, Pizza pizza) {
        
        // Step 1: Logic is calculated by the expert (the Pizza object)
        double price = pizza.calculatePrice();

        // Step 2: Printing is simplified using methods on the objects
        System.out.println("Order for: " + customer.getContactInfo());
        System.out.println("Ship to: " + address); // Uses address.toString()
        System.out.println("Total: $" + price);
    }
}

// Encapsulates customer validation
class Customer {
    private String name;
    private String email;

    public Customer(String name, String email) {
        if (!email.contains("@")) throw new IllegalArgumentException("Invalid email");
        this.name = name;
        this.email = email;
    }
    
    public String getContactInfo() { return name + " (" + email + ")"; }
}

// Encapsulates address formatting
class Address {
    private String street;
    private String city;
    private String zipCode;

    public Address(String street, String city, String zipCode) {
        if (zipCode.length() != 5) throw new IllegalArgumentException("Invalid zip");
        this.street = street;
        this.city = city;
        this.zipCode = zipCode;
    }

    public String toString() { return street + ", " + city + " " + zipCode; }
}

// Encapsulates pricing logic
class Pizza {
    private String size;
    private int toppingsCount;

    public Pizza(String size, int toppingsCount) {
        this.size = size;
        this.toppingsCount = toppingsCount;
    }

    public double calculatePrice() {
        double basePrice = switch (size) {
            case "Small" -> 10.00;
            case "Medium" -> 12.00;
            case "Large" -> 14.00;
            default -> 0.0;
        };
        return basePrice + (toppingsCount * 1.50);
    }
}